models/migrations/v114.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/git"
+
+	"xorm.io/xorm"
+)
+
+func fixPublisherIDforTagReleases(x *xorm.Engine) error {
+
+	type Release struct {
+		ID          int64
+		RepoID      int64
+		Sha1        string
+		TagName     string
+		PublisherID int64
+	}
+
+	type Repository struct {
+		ID      int64
+		OwnerID int64
+		Name    string
+	}
+
+	type User struct {
+		ID    int64
+		Name  string
+		Email string
+	}
+
+	const batchSize = 100
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	var (
+		gitRepoCache = make(map[int64]*git.Repository)
+		gitRepo      *git.Repository
+		repoCache    = make(map[int64]*Repository)
```

Comment:
This could potentially end up holding every repo in the db.

Why don't you query the releases in ascending repo id order then you can get rid of both the gitrepo and repo cache.

In terms of the user cache you could limit it to say 100 users and remove the oldest least releasing user.